### DISCUSSION TITLE
alive? vs finished? vs status

### DISCUSSION THREAD

**Discussion by j-manu** [socketry/async#289] [Q&A] [answered]
`finished?` returns true after self and child tasks have completed execution

`status` can be `:initialized`, `:running`, `:completed`, `:stopped` or `:failed`

1. When is `alive?` false?
2. Completed execution means status is in `:completed`, `:stopped` or `:failed` right?

**Comment by ioquatix** [comment, marked as answer]
Generally speaking, the only interface you should depend on is `wait`. `finished?` relates to whether that part of the task tree has finished executing, including all children.

1. `alive?` is false if the fiber does not exist yet or has completed execution: https://github.com/socketry/async/blob/ffd019d9c1d547926a28fe8f36bf7bfe91d8a168/lib/async/task.rb#L117-L120 note that `@fiber` starts out as `nil`.
2. Yes, that's correct. There is a diagram here that should help: https://socketry.github.io/async/guides/asynchronous-tasks/index#task-lifecycle

**Reply by j-manu** [reply]
Thanks.

> Generally speaking, the only interface you should depend on is wait. finished? relates to whether that part of the task tree has finished executing, including all children.

Yeah but `finished?` is very useful when you have nested async tree and you don't want to wait on the task but periodically check whether the task and all its children have completed.

My use case: When a websocket connection is initiated, an async task with timeout is spawned. The task itself has multiple nested async tasks. The connection handler has a loop which loops until `task.finished?` . The tasks use a an instance of `Thread::Queue` which is read from in the loop and sent via the websocket connection

> `alive?` is false if the fiber does not exist  

When can this happen when using `Async {}` ?

> Yes, that's correct. There is a diagram here that should help: https://socketry.github.io/async/guides/asynchronous-tasks/index#task-lifecycle 

Yeah, I had seen that but wanted to confirm. Would you accept a PR to add `completed_execution?` to task? Right now,
if you want that, you have to do `%i[status completed failed].include?(status)`
